### ISSUE TITLE
Primavera Exception Dates: Off by 1 day?

### ISSUE THREAD

**Issue by dalmm** [joniles/mpxj#33]
I noticed recently a change was made to account for P6 exception dates being days from 12/30/1899 instead of 1900.

I thought this should fix the issue I'm seeing, but it didn't.

I'm using the 7.0.2 NUGET .NET package, and I have to add 1 day to the exceptions after import to make them match what is found in the P6 client.

```
foreach (ProjectCalendarException exception in calendar.CalendarExceptions.ToIEnumerable())
                {
                    if (!exception.Working)
                    {
                        for (var dt = exception.FromDate.ToDateTime(); dt <= exception.ToDate.ToDateTime(); dt = dt.AddDays(1))
                            dates.Add(dt.Date.AddDays(1));
                    }
                }
```

I've tried using the ToDateTime function provided by MPXJ as well as another one I found that does a simple conversion using Year, Month, etc into new DateTime.  Both produce the same results.

When I inspect the Calendar object, the Exception shows, for instance, as 2009/12/31 (-6:00), when it should be 2010/1/1.

**Comment by joniles** [comment]
Can you send me an example XER file?

**Comment by joniles** [comment]
In fact, can you confirm how you are reading the Primavera data (XER or PMXML file, or direct database access)?

**Comment by dalmm** [comment]
Attached is the sample XER file.
[Sample Xer 02.zip](https://github.com/joniles/mpxj/files/1552510/Sample.Xer.02.zip)

For instance, for the XER portion:
(0||0(d|36982)
[Day Calculation](https://www.timeanddate.com/date/dateadded.html?m1=12&d1=30&y1=1899&type=add&ay=&am=&aw=&ad=36982&rec=)

I'm getting:

![image](https://user-images.githubusercontent.com/34243567/33896985-e2030046-df29-11e7-861d-d23b8405a912.png)

**Comment by dalmm** [comment]
I actually have another request, somewhat related to this (exceptions again).
I need to know if the exception has a time range.  For my purposes, it's not necessary to know the exact range (currently), but maybe someone else would need that. Options:
1) include an "isAllDay" flag or something similar
2) include the actual time range

Example (from the same XER file):
`(0||0(d|36982)(      (0||0(s|08:00|f|12:00)())`

**Comment by joniles** [comment]
Based on your observations I took a look at the code which reads calendar exceptions from XER files. With regard to your second request, I found that the code to read "working day exceptions" (calendar exceptions which represent working days) was completely missing! I've put this in place, and it looks like everything is now working as expected.

For the "off by one day" problem with exception dates, this was cause by me not handling timezones correctly when performing the date arithmetic... I've fixed this too. I will be releasing version 7.1.0 shortly which will contain these fixes.

Let me know if the new version works for you!
